client/public/js/frame.js

```diff
@@ -1,41 +1,29 @@
 (function() {
-  function evalSources(sources) {
-    var modules = {};
-    function dirname(str) {
-      var ix = str.lastIndexOf("/");
-      return ix < 0 ? "" : str.slice(0, ix);
-    }
-    function resolvePath(a, b) {
-      if (b[0] === "." && b[1] === "/") {
-        return dirname(a) + b.slice(1);
-      }
-      if (b[0] === "." && b[1] === "." && b[2] === "/") {
-        return dirname(dirname(a)) + b.slice(2);
-      }
-      return b;
-    }
-    return function load(name) {
-      if (modules[name]) {
-        return modules[name].exports;
-      }
-      function require(path) {
-        return load(resolvePath(name, path));
-      }
-      var module = modules[name] = { exports: {} };
-      new Function("module", "exports", "require", sources[name])(module, module.exports, require);
-      return module.exports;
-    };
-  }
-
   var parent;
 
+  // Ensures `main` is only called once.
+  this.window.runCount = 0;
+
   document.addEventListener("DOMContentLoaded", function() {
     window.addEventListener("message", function(event) {
       parent = event.source;
       parent.postMessage("trypurescript", "*");
-      var file = evalSources(event.data)("<file>");
-      if (file.main && typeof file.main === "function") {
-        file.main();
+      const code = `
+      ${event.data.code}
+      if (window.runCount === 0) {
```

Comment:
I see that you attach it to window to reference within this script. Can you explain how main can get invoked more than once?

Reply:
Also, it looks like `main` is invoked infinitely only when a shim is used. Disabling this check and running the `Loops` example doesn't trigger the infinite invoking. Disabling the check and running something that uses `Data.Decimal.fromString` causes `main` to be invoked infinitely.

Reply:
It would be helpful to understand why. For example, what if the user's module has an unsafePerformEffect? Will that get invoked infinitely? Without knowing what's actually happening, I'm not sure that guarding on main this is strong enough. It feels like it's masking another problem to me. How are you verifying that it's being invoked infinitely? Is it getting invoked because the main window keeps trying to update the frame due to some exception and it not being told to stop?

Reply:
Good points!

I think the fastest way to troubleshoot this would be to open an issue on the `es-module-shims` repo and ask the author.

Reply:
I tried to create a reproducible issue and failed to do so. However, I did a large hint.

The code before the below change is made is invoked only once. The changed code is invoked infinitely.
```diff
module Main where

import Prelude

import Effect (Effect)
import Effect.Console (log)
import Data.Decimal as D
+import TryPureScript as TP

main :: Effect Unit
-main = do
+main = TP.render =<< TP.withConsole do
  log $ show $ D.fromString "4232.424242"
```

In the TryPureScript module's FFI for [`withConsoleImpl`](https://github.com/purescript/trypurescript/blob/master/staging/src/TryPureScript.js#L9-L37), we temporarily override `console.log`, `console.error` and `window.onerror`. I'm not sure how that causes problems yet, but that seems to be the issue.

Reply:
It's probably window.onerror, and it's probably because that's how esmodule shims knows when to kick in with an import map (because the import fails). I bet if you don't override window.onerror it will work.

Reply:
Good guess, but that's not it either. The root cause is that `render` calls `document.body.innerHtml += html;`. Disabling that line while enabling the `withConsoleImpl` `log`/`error`/`onerror` handlers reproduces this issue.

I guess modifying the DOM triggers an event that causes `es-module-shims` to reactivate. That also explains why I can add a `script` element after the page loads and it still runs that code.

Reply:
We should probably be rendering into a specific element rather than document.body.

Reply:
Fixed it in latest commit. Thoughts on my solution?

Reply:
Appending the html to the `main` element works without the previous workaround.